### PR TITLE
Add a synthetic NLP dataset

### PR DESCRIPTION
**Pull request by moinnadeem**
(testing GitHub actions tests)

### CODE REVIEW

File: composer/datasets/synthetic.py [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+import random
```
Comment (ravi-mosaicml) [inline composer/datasets/synthetic.py:3]: The synthetic datasets file is getting kidna large...perhaps can you make it into a folder, and then have a `synthetic_batch_pair_dataset.py`, `synthetic_pil_image_dataset.py`, `synthetic_bert_dataset.py`, and `synthetic_lm_dataset.py`?

File: composer/datasets/synthetic.py [hunk 2]
```diff
@@ -21,6 +29,88 @@ class SyntheticDataLabelType(StringEnum):
...
+        tokenizer.normalizer = tokenizers.normalizers.NFKC()
+        tokenizer.pre_tokenizer = tokenizers.pre_tokenizers.ByteLevel()
+        tokenizer.decoder = tokenizers.decoders.ByteLevel()
+        trainer = tokenizers.trainers.WordPieceTrainer(
+            vocab_size=vocab_size,
+            initial_alphabet=tokenizers.pre_tokenizers.ByteLevel.alphabet(),
+            special_tokens=["[PAD]", "[UNK]", "[SEP]", "[CLS]", "[MASK]"],
+        )
+        tokenizer.train_from_iterator(dataset, trainer=trainer)
+        tmp_tokenizer_file = NamedTemporaryFile()
```
Comment (ravi-mosaicml) [inline composer/datasets/synthetic.py:53]: Can you use a context manager for the temporary file, and then call `super().__init__` within this context manager, so the tempfile will automatically disappear once its no longer needed?

File: composer/datasets/synthetic.py [hunk 3]
```diff
@@ -21,6 +29,88 @@ class SyntheticDataLabelType(StringEnum):
...
+            tmp_tokenizer_file.write(f"{token}\n".encode())
+        tmp_tokenizer_file.flush()
+
+        super().__init__(tmp_tokenizer_file.name)
+
+
+class SyntheticHFDataset:
+    """Creates a synthetic HF dataset and passes it to the preprocessing scripts."""
+
+    def __init__(self, num_samples, chars_per_sample, column_names):
```
Comment (ravi-mosaicml) [inline composer/datasets/synthetic.py:64]: type annotations

File: composer/datasets/synthetic.py [hunk 4]
```diff
@@ -21,6 +29,88 @@ class SyntheticDataLabelType(StringEnum):
     CLASSIFICATION_ONE_HOT = "classification_one_hot"
 
 
+class SyntheticBertTokenizer(BertTokenizer):
+
+    def __init__(self, dataset, vocab_size=256):
```
Comment (ravi-mosaicml) [inline composer/datasets/synthetic.py:34]: type annotations

File: composer/datasets/synthetic.py [hunk 5]
```diff
@@ -21,6 +29,88 @@ class SyntheticDataLabelType(StringEnum):
...
+        data = {}
+        for column_name in self.column_names:
+            data[column_name] = [self.generate_sample() for _ in range(self.num_samples)]
+        data['idx'] = list(range(self.num_samples))
+
+        hf_synthetic_dataset = datasets.Dataset.from_dict(data)
+        return hf_synthetic_dataset
+
+    def generate_sample(self):
+        MIN_WORD_LENGTH = 3
```
Comment (ravi-mosaicml) [inline composer/datasets/synthetic.py:88]: can you define constants at the top of the file? or make them hparams?

**Review by ravi-mosaicml — APPROVED** [review]
LGTM; See comments

**Comment by moinnadeem** [comment]
Heads up, this PR isn't actually ready for review yet -- I'll be sure to address your comments though, but this isn't the final version just yet

File: composer/datasets/lm_datasets.py [hunk 6]
```diff
@@ -85,13 +89,36 @@ def initialize_object(self, batch_size: int, dataloader_hparams: DataloaderHpara
...
+            lm_datasets = SyntheticHFDataset(num_samples=self.synthetic_num_unique_samples,
+                                             chars_per_sample=self.train_sequence_length,
+                                             column_names=column_names)
+            lm_datasets = lm_datasets.generate_dataset()
+
+            # flatten the columnar dataset into one column
+            flattened_dataset = functools.reduce(operator.iconcat, [lm_datasets[i] for i in column_names], [])
+            self.tokenizer = SyntheticBertTokenizer(flattened_dataset)
+
+            lm_datasets = lm_datasets.map(lambda inp: self.tokenizer(
```
Comment (hanlint) [inline composer/datasets/lm_datasets.py:105]: this indentation looks odd and difficult to read

**Review by ravi-mosaicml — CHANGES_REQUESTED** [review]
Unapproving per offline discussion

File: composer/datasets/lm_datasets.py [hunk 7]
```diff
@@ -85,13 +89,36 @@ def initialize_object(self, batch_size: int, dataloader_hparams: DataloaderHpara
...
+                text=inp[column_names[0]
```
Comment (hanlint) [inline composer/datasets/lm_datasets.py:106]: this won't handle column_names > 1? or is the (unenforced) convention that `column_names[0]` should always be the text?

File: composer/datasets/synthetic.py [hunk 8]
```diff
@@ -21,6 +29,88 @@ class SyntheticDataLabelType(StringEnum):
...
+
+    def __init__(self, dataset, vocab_size=256):
+        try:
+            import tokenizers
+        except ImportError as e:
+            raise ImportError(
+                'Composer was installed without NLP support. To use NLP with Composer, run: `pip install mosaicml[nlp]`.'
+            ) from e
+
+        tokenizer = tokenizers.Tokenizer(tokenizers.models.WordPiece())
```
Comment (hanlint) [inline composer/datasets/synthetic.py:42]: is it possible to set the `normalizer`, `decoder`, etc. in the `Tokenizer` initialization function itself, rather than in subsequient lines?

File: composer/datasets/synthetic.py [hunk 9]
```diff
@@ -21,6 +29,88 @@ class SyntheticDataLabelType(StringEnum):
...
+        tokenizer.enable_padding(direction="right", pad_id=0, pad_type_id=0, pad_token="[PAD]", pad_to_multiple_of=8)
+        tokenizer.normalizer = tokenizers.normalizers.NFKC()
+        tokenizer.pre_tokenizer = tokenizers.pre_tokenizers.ByteLevel()
+        tokenizer.decoder = tokenizers.decoders.ByteLevel()
+        trainer = tokenizers.trainers.WordPieceTrainer(
```
Comment (hanlint) [inline composer/datasets/synthetic.py:47]: trainer is misleading.. we have our `composer.trainer` as well.
